sys/mips/rootdev.c

```diff
@@ -24,15 +24,35 @@ static void rootdev_unmask_irq(intr_event_t *ie) {
   mips32_bs_c0(C0_STATUS, SR_IM0 << irq);
 }
 
-static void rootdev_intr_setup(device_t *dev, unsigned irq,
-                               intr_handler_t *handler) {
+/* clang-format off */
+static const char *rootdev_intr_name[MIPS_NIRQ] = {
+  [MIPS_SWINT0] = "swint(0)",
+  [MIPS_SWINT1] = "swint(1)",
+  [MIPS_HWINT0] = "hwint(0)",
+  [MIPS_HWINT1] = "hwint(1)",
+  [MIPS_HWINT2] = "hwint(2)",
+  [MIPS_HWINT3] = "hwint(3)",
+  [MIPS_HWINT4] = "hwint(4)",
+  [MIPS_HWINT5] = "hwint(5)",
+};
+/* clang-format on */
+
+static void rootdev_intr_setup(device_t *dev, resource_t *r,
+                               ih_filter_t *filter, ih_service_t *service,
+                               void *arg, const char *name) {
   rootdev_t *rd = dev->parent->state;
-  intr_event_t *event = &rd->intr_event[irq];
-  intr_event_add_handler(event, handler);
+  int irq = r->r_start;
+
+  if (rd->intr_event[irq] == NULL)
+    rd->intr_event[irq] = intr_event_create(
+      dev, irq, rootdev_mask_irq, rootdev_unmask_irq, rootdev_intr_name[irq]);
+
+  r->r_handler =
+    intr_event_add_handler(rd->intr_event[irq], filter, service, arg, name);
 }
 
-static void rootdev_intr_teardown(device_t *dev, intr_handler_t *handler) {
-  intr_event_remove_handler(handler);
+static void rootdev_intr_teardown(device_t *dev, resource_t *irq) {
+  intr_event_remove_handler(irq->r_handler);
```

Comment:
We are leaking an interrupt handler here.

Reply:
We're calling `kfree` in `intr_event_remove_handler`. Done!